src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+    const Shape& past_lens_shape,         // e.g. {batch_seq}
+    const int32_t* subsequence_begins,    // [batch_seq + 1]: start indices of new tokens per sequence
+    const int32_t* block_indices,         // [num_blocks]: block table for each sequence
+    const int32_t* block_indices_begins,  // [batch_seq + 1]: indices into block_indices per sequence
+    const T* scale_ptr,                   // (Optional) attention scale factor (can be nullptr; default = 1)
+    const int32_t* sliding_window_ptr,    // (Optional) sliding window parameter (can be nullptr; default = 0)
+    const T* alibi_slopes,  // (Optional) [num_kv_heads]: per-head bias slopes (can be nullptr; default = 0)
+    const int32_t* max_context_len_ptr,  // max context length (for score output indexing)
+
+    // Rotation parameters (if any is nullptr, rotation is skipped)
+    const int32_t* rotated_block_indices,  // (Optional) [num_rotated_blocks]: blocks to which RoPE is applied
+    const int32_t* rotation_deltas,  // (Optional) [num_rotated_blocks, block_size || 1]: indices into the trig LUT
+    const T* rotation_trig_lut,      // (Optional) LUT: [lut_rows, head_size] (first half: cosines, second half: sines)
+    const Shape& rotated_block_indices_shape,  // shape of rotated_block_indices (e.g. {num_rotated_blocks})
+    const Shape& rotation_deltas_shape,        // shape of rotation_deltas (e.g. {num_rotated_blocks, block_size} or
+                                               // {num_rotated_blocks, 1})
+    const Shape& rotation_trig_lut_shape) {    // shape of rotation_trig_lut
+
+    // Use default values for optional parameters if pointers are null.
+    T scale = scale_ptr ? scale_ptr[0] : T(1);
+    int32_t sliding_window = sliding_window_ptr ? sliding_window_ptr[0] : 0;
+    int32_t max_context_len = max_context_len_ptr ? max_context_len_ptr[0] : 0;
+
+    // Determine dimensions.
+    int32_t num_blocks = kv_cache_shape[0];
+    int32_t num_kv_heads = kv_cache_shape[1];
+    int32_t block_size = kv_cache_shape[2];
+    int32_t head_size = kv_cache_shape[3];
+
+    int32_t batch_tokens = q_shape[0];
+    int32_t query_features = q_shape[1];  // equals num_heads * head_size.
+    int32_t num_heads = query_features / head_size;
+    int32_t batch_seq = !past_lens_shape.empty() ? past_lens_shape[0] : 1;
+
+    // Process each query token.
+    for (int32_t token_idx = 0; token_idx < batch_tokens; token_idx++) {
+        // Determine sequence index using subsequence_begins if available.
+        int32_t seq_idx = 0;
+        if (batch_seq > 1 && subsequence_begins) {
+            for (int32_t s = 0; s < batch_seq; s++) {
+                if (token_idx >= subsequence_begins[s] && token_idx < subsequence_begins[s + 1]) {
+                    seq_idx = s;
+                    break;
+                }
+            }
+        }
+        // Process each head.
+        for (int32_t h = 0; h < num_heads; h++) {
```

Comment:
Did you skip the step of copying key and value to kvcache before the computation like CPU plugin?
https://github.com/openvinotoolkit/openvino/blob/f671bfc54e018c843fbfb5a12fb0b34c7a9a87b1/src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp#L2710-L2712

Reply:
Right, added - now, if the token belongs to the “new” portion of the sequence (as indicated by the subsequence indices), its key and value vectors are copied into the corresponding key/value cache locations